File: teaching/anthropic/anthropic-protocol-book/code/python/ch06/tool_search.py
"""
Chapter 6: Embedding-based Tool Search.

Implements semantic tool discovery using cosine similarity over
embedding vectors. Supports on-demand tool loading for large tool
libraries, matching the behavior of Anthropic's Tool Search Tool
with custom embedding-based search.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional, Tuple
import re as _re


# ---------------------------------------------------------------------------
# Data types
# ---------------------------------------------------------------------------


@dataclass
class ToolSearchEntry:
    """A single tool in the search index."""

    name: str
    description: str
    input_schema: Dict[str, Any]

    def to_search_text(self) -> str:
        """Build the full-text representation used for embedding generation.

        Concatenates the tool name, description, and all parameter
        names plus their descriptions. This provides the search engine
        with a rich surface for semantic matching.
        """
        parts: List[str] = [self.name, self.description]

        props = self.input_schema.get("properties", {})
        for prop_name, prop_def in props.items():
            parts.append(prop_name)
            if isinstance(prop_def, dict):
                desc = prop_def.get("description", "")
                if desc:
                    parts.append(desc)

        return " ".join(parts)


# ---------------------------------------------------------------------------
# Search engine
# ---------------------------------------------------------------------------


class EmbeddingToolSearch:
    """Semantic tool search using embedding-based cosine similarity.

    Designed to mirror Anthropic's Tool Search Tool behavior --
    discover tools on-demand rather than loading all definitions
    into the context window.

    Supports two search modes:
    1. **Semantic**: cosine similarity over hash-based embedding vectors.
    2. **Regex**: direct pattern matching against tool metadata.

    Usage::

        search = EmbeddingToolSearch(embedding_dim=384)
        search.index_tools([
            {"name": "get_weather", "description": "...", "input_schema": {...}},
            {"name": "get_time", "description": "...", "input_schema": {...}},
        ])
        results = search.search("weather forecast", top_k=3)
    """

    def __init__(self, embedding_dim: int = 384) -> None:
        """Initialize the search engine.

        Args:
            embedding_dim: Dimensionality of the embedding vectors.
                Default 384 matches MiniLM-L6-v2. Use 768 for larger models.
        """
        if embedding_dim <= 0:
            raise ValueError("embedding_dim must be positive")
        self._tools: List[ToolSearchEntry] = []
        self._embeddings: Optional[List[List[float]]] = None
        self._embedding_dim = embedding_dim

    # ------------------------------------------------------------------
    # Indexing
    # ------------------------------------------------------------------

    def index_tools(self, tools: List[Dict[str, Any]]) -> None:
        """Build the search index from tool definitions.

        Generates embedding vectors for each tool's search text.
        In production, replace ``_compute_embeddings`` with a call
        to a real embedding model (e.g., ``text-embedding-3-small``,
        ``MiniLM-L6-v2``).

        Args:
            tools: List of tool dicts with ``name``, ``description``,
                   and ``input_schema`` keys.
        """
        self._tools.clear()

        for tool in tools:
            entry = ToolSearchEntry(
                name=tool["name"],
                description=tool.get("description", ""),
                input_schema=tool.get(
                    "input_schema", {"type": "object", "properties": {}}
                ),
            )
            self._tools.append(entry)

        # Generate embeddings for all tools
        search_texts = [t.to_search_text() for t in self._tools]
        self._embeddings = self._compute_embeddings(search_texts)

    def _compute_embeddings(self, texts: List[str]) -> List[List[float]]:
        """Compute embedding vectors for a list of texts.

        Uses a hash-based frequency vector approach for deterministic
        testing. In production, this would be replaced by a call to
        a real embedding model API (e.g., Voyage AI, OpenAI embeddings).

        The hash-based approach ensures:
        - Identical texts produce identical vectors
        - Similar texts produce similar vectors (via shared tokens)
        - Vectors are normalized to unit length for cosine similarity
        """
        dim = self._embedding_dim
        embeddings: List[List[float]] = []

        for text in texts:
            tokens = text.lower().split()
            vec = [0.0] * dim

            for token in tokens:
                idx = hash(token) % dim
                vec[idx] += 1.0

            # L2 normalize
            norm = sum(v * v for v in vec) ** 0.5
            if norm > 0:
                vec = [v / norm for v in vec]

            embeddings.append(vec)

        return embeddings

    # ------------------------------------------------------------------
    # Search
    # ------------------------------------------------------------------

    def search(self, query: str, top_k: int = 5) -> List[Dict[str, Any]]:
        """Search for tools relevant to a natural-language query.

        Generates an embedding for the query and computes cosine
        similarity against all indexed tool embeddings.

        Args:
            query: Natural language description of the desired capability.
            top_k: Maximum number of results to return.

        Returns:
            List of result dicts sorted by descending score, each containing
            ``name``, ``description``, ``input_schema``, and ``score`` (0-1).
        """
        if not self._tools or self._embeddings is None:
            return []

        query_embedding = self._compute_embeddings([query])[0]

        # Compute pairwise similarities
        scored: List[Tuple[int, float]] = []
        for i, emb in enumerate(self._embeddings):
            sim = self._cosine_similarity(query_embedding, emb)
            scored.append((i, sim))

        # Sort by score descending
        scored.sort(key=lambda x: x[1], reverse=True)

        results: List[Dict[str, Any]] = []
        for idx, score in scored[:top_k]:
            if score > 0:
                tool = self._tools[idx]
                results.append(
                    {
                        "name": tool.name,
                        "description": tool.description,
                        "input_schema": tool.input_schema,
                        "score": round(score, 4),
                    }
                )

        return results

    def search_by_regex(self, pattern: str) -> List[Dict[str, Any]]:
        """Search for tools using Python regex (mirrors the Regex variant
        of Anthropic's Tool Search Tool).

        Matches the pattern against the full search text of each tool
        (name + description + parameter names + parameter descriptions).

        Args:
            pattern: Python regex pattern (e.g., ``"(?i)weather"``).

        Returns:
            Matching tool definitions.
        """
        results: List[Dict[str, Any]] = []
        for tool in self._tools:
            search_text = tool.to_search_text()
            if _re.search(pattern, search_text):
                results.append(
                    {
                        "name": tool.name,
                        "description": tool.description,
                        "input_schema": tool.input_schema,
                    }
                )
        return results

    # ------------------------------------------------------------------
    # Utilities
    # ------------------------------------------------------------------

    @staticmethod
    def _cosine_similarity(a: List[float], b: List[float]) -> float:
        """Compute cosine similarity between two vectors.

        Returns a value between -1 and 1, where 1 means identical
        direction and 0 means orthogonal.
        """
        if len(a) != len(b):
            raise ValueError(
                f"Vector dimension mismatch: {len(a)} vs {len(b)}"
            )

        dot_product = sum(x * y for x, y in zip(a, b))
        norm_a = sum(x * x for x in a) ** 0.5
        norm_b = sum(x * x for x in b) ** 0.5

        if norm_a == 0.0 or norm_b == 0.0:
            return 0.0

        return dot_product / (norm_a * norm_b)

    # ------------------------------------------------------------------
    # Properties
    # ------------------------------------------------------------------

    @property
    def tool_count(self) -> int:
        """Number of tools in the index."""
        return len(self._tools)

    @property
    def embedding_dim(self) -> int:
        """Current embedding dimension."""
        return self._embedding_dim

    @property
    def is_indexed(self) -> bool:
        """Whether the index has been built."""
        return self._embeddings is not None and len(self._tools) > 0

File: teaching/anthropic/anthropic-protocol-book/code/node/ch06/ToolSearch.ts
/**
 * Chapter 6: Embedding-based Tool Search (TypeScript).
 *
 * Semantic tool discovery using cosine similarity over hash-based
 * embedding vectors. Mirrors Anthropic's Tool Search Tool behavior
 * with custom embedding-based search.
 */

interface ToolSearchEntryData {
  name: string;
  description: string;
  input_schema: Record<string, unknown>;
}

interface SearchResult {
  name: string;
  description: string;
  input_schema: Record<string, unknown>;
  score: number;
}

/**
 * Semantic tool search using embedding-based cosine similarity.
 *
 * Usage:
 * ```typescript
 * const search = new EmbeddingToolSearch(384);
 * search.indexTools([...toolDefinitions]);
 * const results = search.search("weather forecast", 3);
 * ```
 */
export class EmbeddingToolSearch {
  private tools: ToolSearchEntryData[] = [];
  private embeddings: number[][] | null = null;
  private readonly dim: number;

  constructor(embeddingDim: number = 384) {
    if (embeddingDim <= 0) {
      throw new Error("embeddingDim must be positive");
    }
    this.dim = embeddingDim;
  }

  // ------------------------------------------------------------------
  // Indexing
  // ------------------------------------------------------------------

  indexTools(
    tools: Array<{
      name: string;
      description: string;
      input_schema: Record<string, unknown>;
    }>,
  ): void {
    this.tools = tools.map((t) => ({
      name: t.name,
      description: t.description ?? "",
      input_schema: t.input_schema ?? { type: "object", properties: {} },
    }));

    const texts = this.tools.map((t) => this.toSearchText(t));
    this.embeddings = this.computeEmbeddings(texts);
  }

  private toSearchText(entry: ToolSearchEntryData): string {
    const parts: string[] = [entry.name, entry.description];
    const props = (entry.input_schema as Record<string, unknown>)
      .properties as Record<string, Record<string, unknown>> | undefined;
    if (props) {
      for (const [propName, propDef] of Object.entries(props)) {
        parts.push(propName);
        if (propDef && typeof propDef.description === "string") {
          parts.push(propDef.description);
        }
      }
    }
    return parts.join(" ");
  }

  private computeEmbeddings(texts: string[]): number[][] {
    return texts.map((text) => {
      const tokens = text.toLowerCase().split(/\s+/);
      const vec = new Array<number>(this.dim).fill(0);
      for (const token of tokens) {
        const idx = this.hashToken(token) % this.dim;
        vec[idx] += 1.0;
      }
      // L2 normalize
      const norm = Math.sqrt(vec.reduce((s, v) => s + v * v, 0));
      if (norm > 0) {
        for (let i = 0; i < vec.length; i++) {
          vec[i] /= norm;
        }
      }
      return vec;
    });
  }

  private hashToken(token: string): number {
    let hash = 0;
    for (let i = 0; i < token.length; i++) {
      const char = token.charCodeAt(i);
      hash = ((hash << 5) - hash + char) | 0;
    }
    return Math.abs(hash);
  }

  // ------------------------------------------------------------------
  // Search
  // ------------------------------------------------------------------

  search(query: string, topK: number = 5): SearchResult[] {
    if (this.tools.length === 0 || !this.embeddings) {
      return [];
    }

    const queryEmb = this.computeEmbeddings([query])[0]!;
    const scores: Array<{ index: number; score: number }> = [];

    for (let i = 0; i < this.embeddings.length; i++) {
      scores.push({
        index: i,
        score: EmbeddingToolSearch.cosineSimilarity(
          queryEmb,
          this.embeddings[i]!,
        ),
      });
    }

    scores.sort((a, b) => b.score - a.score);

    return scores.slice(0, topK).map(({ index, score }) => ({
      name: this.tools[index]!.name,
      description: this.tools[index]!.description,
      input_schema: this.tools[index]!.input_schema,
      score: Math.round(score * 10000) / 10000,
    }));
  }

  searchByRegex(
    pattern: string,
  ): Array<{
    name: string;
    description: string;
    input_schema: Record<string, unknown>;
  }> {
    const regex = new RegExp(pattern);
    return this.tools
      .filter((t) => regex.test(this.toSearchText(t)))
      .map((t) => ({
        name: t.name,
        description: t.description,
        input_schema: t.input_schema,
      }));
  }

  // ------------------------------------------------------------------
  // Utilities
  // ------------------------------------------------------------------

  static cosineSimilarity(a: number[], b: number[]): number {
    if (a.length !== b.length) {
      throw new Error(
        `Vector dimension mismatch: ${a.length} vs ${b.length}`,
      );
    }
    let dot = 0;
    let normA = 0;
    let normB = 0;
    for (let i = 0; i < a.length; i++) {
      dot += a[i]! * b[i]!;
      normA += a[i]! * a[i]!;
      normB += b[i]! * b[i]!;
    }
    if (normA === 0 || normB === 0) return 0;
    return dot / (Math.sqrt(normA) * Math.sqrt(normB));
  }

  // ------------------------------------------------------------------
  // Properties
  // ------------------------------------------------------------------

  getToolCount(): number {
    return this.tools.length;
  }

  getEmbeddingDim(): number {
    return this.dim;
  }

  get isIndexed(): boolean {
    return this.embeddings !== null && this.tools.length > 0;
  }
}
